README
README for pgtools

The pgtools project is a collection of perl tools written to help obtain
information about various entities of a PostgreSQL server database via
the command line. It includes individual programs that perform seperate tasks. 


ABOUT
The following tools are included in this package:

 pgcol.pl: displays all tables of a specific database that contain a specific column.

pgcomm.pl: queries a column's comment field to display the purpose of the column.

pglist.pl: looks up the last modification time of a PostgreSQL database and a respective table.

 pgoid.pl: looks up the oid for a PostgreSQL database or table.

 pgtab.pl: displays specific and often accessed pre-configured tables.



REQUIREMENTS
Perl 5.0.8 or up


CONFIGURATION
Each program included in this package has various configuration
sections that will need to be configured. They are identified
in each program.


INSTALLATION
see INSTALL file


TODO    - see TODO file
LICENSE - see LICENSE file


ChangeLog
ChangeLog.txt (pgutil)
===========================
v0.1.0
------
initial release


pgcol.pl
#!/usr/bin/env perl
#############################################################################
# PROGRAM
#     pgcol.pl
#
# SYNOPSIS
#     This program displays all tables of a specific database that contain a 
#     specific column.
#
# USAGE
#     pgcol.pl [OPTIONS]
#
#     Options:
#         -c [column]         Column name.
#
#         -d [database]       Database name.
#
#         -h                  Usage
#
# MODULES/FILES REQUIRED
#     DBI
#     Time::localtime
#     Data::Dump
#     Getopt::Std
#
# PROGRAMS REQUIRED
#     None
#
# AUTHOR
#     Steve Dickinson
#
# COMMENTS
#############################################################################
# INTERRUPT SIGNALS
#############################################################################
$SIG{'INT' } = 'clean_up'; $SIG{'HUP' } = 'clean_up';
$SIG{'QUIT'} = 'clean_up'; $SIG{'TRAP'} = 'clean_up';
$SIG{'ABRT'} = 'clean_up'; $SIG{'STOP'} = 'clean_up';
$SIG{'TSTP'} = 'clean_up';

use strict;
use warnings;

##############################################################################
# Configure the following variables as necessary for your local system.
# No other configuration is necessary.
##############################################################################
my $driver		= "dbi:Pg";				# DBI data_source driver
my $username 	= "pgsql";				# database user
my $password 	= "";					# database user password
my $default_db	= "template1";			# default database used for connection 
##############################################################################

use DBI;
use Time::localtime;
use Data::Dump qw(dump);
use Getopt::Std;
$Getopt::Std::STANDARD_HELP_VERSION = 1;

my $VERSION = "0.1";

my (%opts, $prog, $data_source, $arg_column, $arg_database, $query, $fields, $field1, $field2, $name, $results);
$prog = $0;

sub VERSION_MESSAGE {
	print $prog . " v." . $VERSION;
	exit(0);
}

sub HELP_MESSAGE {
my $usage = <<END;
NAME
pgcol - Display PostgreSQL tables and views that refer to a specific column

SYNOPSIS
    pgcol -d [database_name] -c [column_name]

DESCRIPTION
    This program displays all tables and views of a specific database that 
    refer to a specific column.

    The following options are available:

    -c [column]         Column name.

    -d [database]       Database name.

    -h                  Usage


Syntax examples:
> pgcol.pl -d college -c countyid
classroom
classroom_v
book
book_v
computer
computer__v
location
location_v

END

	print $usage;
	exit(0);
}

my $opts_status = getopts('c:d:h', \%opts);
if (!$opts_status) {
	print "\n";
	HELP_MESSAGE();
}

if (defined($opts{h})) {
	HELP_MESSAGE();
}

if (defined($opts{c})) {
	$arg_column = $opts{c};
	$arg_column = sanitize($arg_column);
}

if (!defined($arg_column)) {
    print "\nError: a table column name is required.\n\n";
    HELP_MESSAGE();
}

if (defined($opts{d})) {
	$arg_database = $opts{d};
	$arg_database = sanitize($arg_database);
}

if (!defined($arg_database)) {
	$arg_database = $default_db;
}

my $column_name	= "table_name";
$data_source  	= "$driver:dbname=$arg_database";
$query			= "SELECT $column_name FROM information_schema.columns WHERE column_name = '$arg_column' ORDER BY table_name";
get_results($data_source, $query);

exit(0);

sub get_results {
	my ($this_data_source, $this_query) = @_;
	my ($dbh, $sth, $query_results, $error);

	my %attr = (
    	PrintError => 0,
    	RaiseError => 1,
	);

	$dbh 			= DBI->connect($this_data_source, $username, $password, \%attr) or die $DBI::errstr;
	$sth 			= $dbh->prepare($this_query);
	$sth->execute();
	$query_results 	= $sth->fetchall_arrayref({});
	$sth->finish;
	$dbh->disconnect;

	foreach my $row (@$query_results) {
		print $row->{$column_name} . "\n";
	}
}

sub sanitize {
	my ($input) = @_;

	$input =~s /\s+//g;
	$input =~s /;//g;
	$input =~s /[^\w]+//g;

	return $input;
}

sub clean_up {
    exit(0);
}


pgoid.pl
#!/usr/bin/env perl
#############################################################################
# PROGRAM
#     pgoid.pl
#
# SYNOPSIS
#     This program looks up the oid for a PostgreSQL database or table.
#     Typically need to run using sudo.
#
# USAGE
#     sudo pgoid.pl [OPTIONS]
#
#     Options:
#         -d [database]       Database name.
#
#         -h                  Usage
#
#         -t [table]          Table name.
#
# MODULES/FILES REQUIRED
#     DBI
#     Time::localtime
#     Data::Dump
#     Getopt::Std
#
# PROGRAMS REQUIRED
#     None
#
# AUTHOR
#     Steve Dickinson
#
# COMMENTS
#############################################################################
# INTERRUPT SIGNALS
#############################################################################
$SIG{'INT' } = 'clean_up'; $SIG{'HUP' } = 'clean_up';
$SIG{'QUIT'} = 'clean_up'; $SIG{'TRAP'} = 'clean_up';
$SIG{'ABRT'} = 'clean_up'; $SIG{'STOP'} = 'clean_up';
$SIG{'TSTP'} = 'clean_up';

use strict;
use warnings;

##############################################################################
# Configure the following variables as necessary for your local system.
# No other configuration is necessary.
# *NOTE: You will more than likely need to run this program in 'sudo' mode
#        to access the secured files it needs to look up.
##############################################################################
my $driver		= "dbi:Pg";				# DBI data_source driver
my $username 	= "pgsql";				# database user
my $password 	= "";					# database user password
my $pgdata_path	= "/pg92/data/base";	# $PGDATA location (pgXX part will change w/ Postgres version)
my $default_db	= "template1";			# default database used for connection 
##############################################################################

use DBI;
use Time::localtime;
use Data::Dump qw(dump);
use Getopt::Std;
$Getopt::Std::STANDARD_HELP_VERSION = 1;

my $VERSION = "0.1";

my (%opts, $prog, $data_source, $arg_database, $arg_table, $query, $oid, 
	$name, $results, $dbh);
$prog = $0;

$pgdata_path =~ /^\/pg(\d+)\//;
my $pg_version = $1;

sub VERSION_MESSAGE {
	print $prog . " v." . $VERSION;
	exit(0);
}

sub HELP_MESSAGE {
my $usage = <<END;
NAME
pgoid - Get PostgreSQL database or table oid.

SYNOPSIS
     sudo pgoid [-d ] database_name
     sudo pgoid [-d ] database_name [-t] table_name

DESCRIPTION
    This program looks up the oid of a PostgreSQL database or table.
    Typically it needs to be run in sudo mode.

    The following options are available:
    -d [database]       Database name. (required option)

    -h                  Usage

    -t [table]          Table name.


EXAMPLES
> sudo $prog -d college 
1785034

> sudo $prog -d college -t students
1775990

END

	print $usage;
	exit(0);
}

if (! -d $pgdata_path) {
	print "ERROR: \$PGDATA path '$pgdata_path' does not exist!\n";
	print "       Perhaps you should try running this using sudo.\n";
	exit;
} 

my $opts_status = getopts('d:t:h', \%opts);
if (!$opts_status) {
	print "\n";
	HELP_MESSAGE();
}

if (defined($opts{h})) {
	HELP_MESSAGE();
}

if (defined($opts{d})) {
	$arg_database = $opts{d};
	$arg_database = sanitize($arg_database);
}

if (defined($opts{t})) {
	$arg_table = $opts{t};
	$arg_table = sanitize($arg_table);
}

if (!defined($opts{d}) && !defined($opts{t})) {
	print "\nError: a database name or table name is required.\n\n";
	HELP_MESSAGE();
}

if (!defined($opts{d}) && defined($opts{t})) {
	print "\nError: a database name must accompany a table name.\n\n";
	HELP_MESSAGE();
}

format LINE = 
@<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<< @*
$name, $results
.

if (defined($opts{d})) {
	#-rw-------  1 pgsql  pgsql  4 Dec 14 05:00 /pg90/data/base/753229/PG_VERSION
	$name 			= $arg_database;
	$data_source  	= "$driver:dbname=$arg_database";
	$dbh 			= DBI->connect($data_source, $username, $password) or die $DBI::errstr;
	$query			= "SELECT oid FROM pg_database WHERE datname = ?;";
	$oid = get_oid($dbh, $query, $arg_database);
	$dbh->disconnect;
}

if (defined($opts{t})) {
	#-rw-------  1 pgsql  pgsql  25616384 Dec 15 03:39 /pg90/data/base/736895/738844
	$name 			= $arg_database . "." . $arg_table;
	$data_source  	= "$driver:dbname=$arg_database";
	$dbh 			= DBI->connect($data_source, $username, $password) or die $DBI::errstr;
	$query			= "SELECT oid FROM pg_class WHERE relname = ? AND relkind = 'r';";
	$oid = get_oid($dbh, $query, $arg_table);
	$dbh->disconnect;
}

if (!defined($oid)) {
	print "ERROR: OID for $name could not be found!\n";
} else {
	print "$oid\n";
}

exit(0);

sub sanitize {
	my ($input) = @_;

	$input =~s /\s+//g;
	$input =~s /;//g;
	$input =~s /[^\w]+//g;

	return $input;
}

sub get_oid {
	my ($dbh, $this_query, $this_arg) = @_;
	my ($sth, $query_results, $oid);

	$sth 			= $dbh->prepare($this_query);
	$sth->execute($this_arg);
	$query_results 	= $sth->fetchrow_arrayref;
	$sth->finish;

	if ($query_results->[0] && $query_results->[0] =~ /^(\d+)$/) {
		$oid = $1;
	} else {
		$oid = undef;
	}

	return $oid;
}

sub clean_up {
    exit(0);
}


pgtab.pl
#!/usr/bin/env perl
#############################################################################
# PROGRAM
#     pgtab.pl
#
# SYNOPSIS
#     This program displays specific and often accessed pre-configured tables.
#
# USAGE
#     sudo pgtab.pl [OPTIONS]
#
#     Options:
#         -d [database]       Database name.
#
#         -h                  Usage
#
#         -i                  Displays currently configured tables.
#
#         -t [table]          Table name.
#
# MODULES/FILES REQUIRED
#     DBI
#     Time::localtime
#     Data::Dump
#     Getopt::Std
#
# PROGRAMS REQUIRED
#     None
#
# AUTHOR
#     Steve Dickinson
#
# COMMENTS
#############################################################################
# INTERRUPT SIGNALS
#############################################################################
$SIG{'INT' } = 'clean_up'; $SIG{'HUP' } = 'clean_up';
$SIG{'QUIT'} = 'clean_up'; $SIG{'TRAP'} = 'clean_up';
$SIG{'ABRT'} = 'clean_up'; $SIG{'STOP'} = 'clean_up';
$SIG{'TSTP'} = 'clean_up';

use strict;
use warnings;

##############################################################################
# Configure the following variables as necessary for your local system.
# No other configuration is necessary.
##############################################################################
my $driver		= "dbi:Pg";				# DBI data_source driver
my $username 	= "pgsql";				# database user
my $password 	= "";					# database user password
my $default_db	= "template1";			# default database used if not specified 
my $tables		= {                     # tables and fields to display
	"book"				=> "id, name",
	"location"			=> "id, name",
	"student"			=> "id, name",
	"student_status"	=> "id, abbr, oneword",
	"room"				=> "id, nickname, description",
};
##############################################################################

use DBI;
use Time::localtime;
use Data::Dump qw(dump);
use Getopt::Std;
$Getopt::Std::STANDARD_HELP_VERSION = 1;

my $VERSION = "0.1";

my (%opts, $prog, $data_source, $arg_database, $arg_table, $query, $fields, $field1, $field2, $name, $results);
$prog = $0;

sub VERSION_MESSAGE {
	print $prog . " v." . $VERSION;
	exit(0);
}

sub HELP_MESSAGE {
my $usage = <<END;
NAME
pgtab - Displays PostgreSQL table structures of specific user configurable table names.

SYNOPSIS
    pgtab -i OR -d [database_name] -t [table_name] 

DESCRIPTION
    This program displays table structures of specific user configurable table names.

    The following options are available:

    -d [database]       Database name.

    -h                  Usage

    -i                  Displays currently configured tables.

    -t [table]          Table name.


Syntax examples:
> pgtab.pl -d college -t books (all options specified)
> pgtab.pl -t books            (defaults to college database)
> pgtab.pl books               (defaults to college database and considers this a table reference)

END

	print $usage;
	exit(0);
}

sub INFO_MESSAGE {
	my ($table, $fields);
format LIST = 
@<<<<<<<<<<<<<<<<<<<<< @*
$table, $fields
.

	$~ = "LIST";
	print "\nThe following tables are currently configured for display:\n\n";
	$table = "TABLE";
	$fields = "FIELDS";
	write();
	foreach $table (sort keys %{$tables}) {
		$fields =  $tables->{$table};
		write();
	}
	exit(0);
}

my $opts_status = getopts('d:t:hi', \%opts);
if (!$opts_status) {
	print "\n";
	HELP_MESSAGE();
}

if (defined($opts{h})) {
	HELP_MESSAGE();
}

if (defined($opts{i})) {
	INFO_MESSAGE();
}

if (defined($opts{d})) {
	$arg_database = $opts{d};
	$arg_database = sanitize($arg_database);
}

if (defined($opts{t})) {
	$arg_table = $opts{t};
	$arg_table = sanitize($arg_table);
	if (! defined($tables->{$arg_table})) {
		print "\nError: Table '$arg_table' not defined!\n\n";
		exit(1);
	}
} else {
	my $args = $#ARGV + 1;
	if ($args == 1) {
		$arg_table = $ARGV[0];
	} else {
		print "\nError: Incorrect number of arguments specified!\n\n";
		HELP_MESSAGE();
	}
}	
$fields = $tables->{$arg_table};

if (!defined($arg_database)) {
	$arg_database = $default_db;
}

if (!defined($arg_table)) {
	print "\nError: a database table name is required.\n\n";
	HELP_MESSAGE();
} else {
	if (!defined($tables->{$arg_table})) {
		print "\nError: Table $arg_table is not currently configured.\n";
		INFO_MESSAGE();
	}
}

$data_source  	= "$driver:dbname=$arg_database";
$query			= "SELECT $fields FROM $arg_table WHERE seq = 0 ORDER BY id";
get_results($data_source, $query);

exit(0);

sub get_results {
	my ($this_data_source, $this_query) = @_;
	my ($dbh, $sth, $query_results, $error);

	my %attr = (
    	PrintError => 0,
    	RaiseError => 1,
	);

	$dbh 			= DBI->connect($this_data_source, $username, $password, \%attr) or die $DBI::errstr;
	$sth 			= $dbh->prepare($this_query);
	$sth->execute();
	$query_results 	= $sth->fetchall_arrayref;
	$sth->finish;
	$dbh->disconnect;

	print "$fields\n";
	print dump $query_results;
}

sub sanitize {
	my ($input) = @_;

	$input =~s /\s+//g;
	$input =~s /;//g;
	$input =~s /[^\w]+//g;

	return $input;
}

sub clean_up {
    exit(0);
}


pgcomm.pl
#!/usr/bin/env perl
#############################################################################
# PROGRAM
#     pgcomm.pl
#
# SYNOPSIS
#     This program queries a column's comment field to display the purpose
#     of the column. Typically need to run using sudo.
#
# USAGE
#     sudo pglist.pl [OPTIONS]
#
#     Options:
#         -c [column]         Column name.
#
#         -d [database]       Database name.
#
#         -h                  Usage
#
#         -t [table]          Table name.
#
# MODULES/FILES REQUIRED
#     DBI
#     Time::localtime
#     Data::Dump
#     Getopt::Std
#
# PROGRAMS REQUIRED
#     None
#
# AUTHOR
#     Steve Dickinson
#
# COMMENTS
#     Example alias:
#     alias pgcm='sudo ~/bin/pgcomm.pl'
#############################################################################
# INTERRUPT SIGNALS
#############################################################################
$SIG{'INT' } = 'clean_up'; $SIG{'HUP' } = 'clean_up';
$SIG{'QUIT'} = 'clean_up'; $SIG{'TRAP'} = 'clean_up';
$SIG{'ABRT'} = 'clean_up'; $SIG{'STOP'} = 'clean_up';
$SIG{'TSTP'} = 'clean_up';

use strict;
use warnings;

##############################################################################
# Configure the following variables as necessary for your local system.
# No other configuration is necessary.
# *NOTE: You will more than likely need to run this program in 'sudo' mode
#        to access the secured files it needs to look up.
##############################################################################
my $driver		= "dbi:Pg";				# DBI data_source driver
my $username 	= "pgsql";				# database user
my $password 	= "";					# database user password
my $pgdata_path	= "/pg92/data/base";	# $PGDATA location (pgXX part will change w/ Postgres version)
##############################################################################

use DBI;
use Time::localtime;
use Data::Dump qw(dump);
use Getopt::Std;
$Getopt::Std::STANDARD_HELP_VERSION = 1;

my $VERSION = "0.1";

my (%opts, $data_source, $arg_database, $arg_table, $arg_column, $query, $db_oid, $tb_oid, 
	$oid, $name, $results, $dbh, $sth, $query_results, $col_num, $comment);

my $batch_mode 	= 0;
my $prog 		= $0;

$pgdata_path =~ /^\/pg(\d+)\//;
my $pg_version = $1;

sub VERSION_MESSAGE {
	print $prog . " v." . $VERSION;
	exit(0);
}

sub HELP_MESSAGE {
my $usage = <<END;
NAME
pgcomm - Display PostgreSQL table's column comment field

SYNOPSIS
     sudo pgcomm -d [database_name] -t [table_name] -c [column_name]

DESCRIPTION
    This program looks up the comment field of a PostgreSQL 
    database table. Typically it needs to be run in sudo mode.

    The following options are available:

    -b                  Batch mode. Used to condense the program's output to one line per execution.

    -c [column]         Column name.

    -d [database]       Database name.

    -h                  Usage

    -t [table]          Table name.


EXAMPLES
> sudo $prog -d college -t students -c ssn
Column description:
This column is used to store the student's encrypted social security number.

END

	print $usage;
	exit(0);
}

if (! -d $pgdata_path) {
	print "ERROR: \$PGDATA path '$pgdata_path' does not exist!\n";
	print "       Perhaps you should try running this using sudo.\n";
	exit;
} 

my $opts_status = getopts('c:d:t:bh', \%opts);
if (!$opts_status) {
	print "\n";
	HELP_MESSAGE();
}

if (defined($opts{h})) {
	HELP_MESSAGE();
}

if (defined($opts{b})) {
	$batch_mode = 1;
}

if (defined($opts{c})) {
	$arg_column = $opts{c};
	$arg_column = sanitize($arg_column);
}

if (defined($opts{d})) {
	$arg_database = $opts{d};
	$arg_database = sanitize($arg_database);
}

if (defined($opts{t})) {
	$arg_table = $opts{t};
	$arg_table = sanitize($arg_table);
}

if (!defined($opts{d}) || !defined($opts{t}) || !defined($opts{c})) {
	print "\nError: A database name, a table name and a column name must be given.\n\n";
	HELP_MESSAGE();
}

format LINE = 
@<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<< @*
$name, $results
.

#-rw-------  1 pgsql  pgsql  25616384 Dec 15 03:39 /pg90/data/base/736895/738844
$data_source  	= "$driver:dbname=$arg_database";
$dbh 			= DBI->connect($data_source, $username, $password) or die $DBI::errstr;

$oid 			= `sudo ./pgoid.pl -d $arg_database -t $arg_table`;
chomp ($oid);

$query			= "SELECT ordinal_position FROM information_schema.columns WHERE table_name = ? AND column_name = ?;";
$sth 			= $dbh->prepare($query);
$sth->execute($arg_table, $arg_column);
$query_results 	= $sth->fetchrow_arrayref;
$col_num 		= $query_results->[0];
if ($col_num !~ /^\d+$/) {
	print "\nERROR: Could not find column '$arg_column'!\n\n";
	exit(1);
}

$query			= "SELECT col_description(?,?)";
$sth 			= $dbh->prepare($query);
$sth->execute($oid, $col_num);
$query_results 	= $sth->fetchrow_arrayref;
$comment = $query_results->[0];
if (!defined($comment)) {
	$comment = "[undefined]";
}
$sth->finish;

if ($batch_mode == 0) {
	print "\n" . $arg_database . "." . $arg_table . "." . $arg_column . "\n";
	print "\nColumn description:\n";
	print "$comment\n";
	print "\n";
} else {
	print $arg_database . "." . $arg_table . "." . $arg_column . ": " . $comment . "\n";
}

$dbh->disconnect;

exit(0);

sub sanitize {
	my ($input) = @_;

	$input =~s /\s+//g;
	$input =~s /;//g;
	$input =~s /[^\w]+//g;

	return $input;
}

sub get_oid {
	my ($type, $query_results) = @_;
	my $oid;

	if ($query_results && $query_results =~ /^(\d+)$/) {
		$oid = $1;
	} else {
		$~ = "LINE";
		$results =  "ERROR: Could not determine oid for $type $name! Does $name exist?\n";
		write();
		exit;
	}

	return $oid;
}

sub clean_up {
    exit(0);
}


pglist.pl
#!/usr/bin/env perl
#############################################################################
# PROGRAM
#     pglist.pl
#
# SYNOPSIS
#     This program looks up the last modification time of a PostgreSQL
#     database and a respective table. Typically need to run using sudo.
#
# USAGE
#     sudo pglist.pl [OPTIONS]
#
#     Options:
#         -a                  If the -d option is specified, show all tables.
#
#         -A                  If the -d option is specified, show only tables that have a primary key.
#
#         -d [database]       Database name.
#
#         -h                  Usage
#
#         -t [table]          Table name.
#
# MODULES/FILES REQUIRED
#     DBI
#     Time::localtime
#     Data::Dump
#     Getopt::Std
#
# PROGRAMS REQUIRED
#     None
#
# AUTHOR
#     Steve Dickinson
#
# COMMENTS
#     Example alias:
#     alias pgls='sudo ~/bin/pglist.pl'
#############################################################################
# INTERRUPT SIGNALS
#############################################################################
$SIG{'INT' } = 'clean_up'; $SIG{'HUP' } = 'clean_up';
$SIG{'QUIT'} = 'clean_up'; $SIG{'TRAP'} = 'clean_up';
$SIG{'ABRT'} = 'clean_up'; $SIG{'STOP'} = 'clean_up';
$SIG{'TSTP'} = 'clean_up';

use strict;
use warnings;

##############################################################################
# Configure the following variables as necessary for your local system.
# No other configuration is necessary.
# *NOTE: You will more than likely need to run this program in 'sudo' mode
#        to access the secured files it needs to look up.
##############################################################################
my $driver		= "dbi:Pg";				# DBI data_source driver
my $username 	= "pgsql";				# database user
my $password 	= "";					# database user password
my $pgdata_path	= "/pg92/data/base";	# $PGDATA location (pgXX part will change w/ Postgres version)
my $default_db	= "template1";			# default database used for connection 
##############################################################################

use DBI;
use Time::localtime;
use Data::Dump qw(dump);
use Getopt::Std;
$Getopt::Std::STANDARD_HELP_VERSION = 1;

my $VERSION = "0.1";

my (%opts, $prog, $data_source, $arg_database, $arg_table, $query, $db_oid, $tb_oid, 
	$arg_all, $name, $results, $dbh);
$prog = $0;

$pgdata_path =~ /^\/pg(\d+)\//;
my $pg_version = $1;

sub VERSION_MESSAGE {
	print $prog . " v." . $VERSION;
	exit(0);
}

sub HELP_MESSAGE {
my $usage = <<END;
NAME
pglist - List PostgreSQL table modification timestamps

SYNOPSIS
     sudo pglist [-d [-a | -A]] database_name
     sudo pglist [-d [-a | -A]] database_name [-t] table_name

DESCRIPTION
    This program looks up the last modification times of a PostgreSQL 
    database and its respective tables. Typically it needs to be run in sudo mode.

    The following options are available:

    -a                  If the -d option is specified, show all tables.

    -A                  If the -d option is specified, show only tables that have a primary key.

    -d [database]       Database name. (required option)

    -h                  Usage

    -t [table]          Table name.


EXAMPLES
> sudo $prog -d college -a 
college                Mon Dec 12 10:53:20 2011 /pg${pg_version}/data/base/555555/PG_VERSION
college.students       Fri Dec 16 03:55:42 2011 /pg${pg_version}/data/base/555555/791234
college.books          Sat Dec 31 10:30:12 2011 /pg${pg_version}/data/base/555555/795555
college.rooms          Sun Dec 18 19:22:41 2011 /pg${pg_version}/data/base/555555/793245
college.classes        Sun Dec 25 17:10:01 2011 /pg${pg_version}/data/base/555555/799999
college.schedules      Thu Dec 29 02:43:05 2011 /pg${pg_version}/data/base/555555/797737

> sudo $prog -d college -t students
college                Mon Dec 12 10:53:20 2011 /pg${pg_version}/data/base/555555/PG_VERSION
college.students       Fri Dec 16 03:55:42 2011 /pg${pg_version}/data/base/555555/791234

END

	print $usage;
	exit(0);
}

if (! -d $pgdata_path) {
	print "ERROR: \$PGDATA path '$pgdata_path' does not exist!\n";
	print "       Perhaps you should try running this using sudo.\n";
	exit;
} 

my $opts_status = getopts('d:t:aAh', \%opts);
if (!$opts_status) {
	print "\n";
	HELP_MESSAGE();
}

if (defined($opts{h})) {
	HELP_MESSAGE();
}

if (defined($opts{d})) {
	$arg_database = $opts{d};
	$arg_database = sanitize($arg_database);
}

if (defined($opts{t})) {
	$arg_table = $opts{t};
	$arg_table = sanitize($arg_table);
}

if (defined($opts{a})) {
	$arg_all = "true";
}

if (defined($opts{A})) {
	$arg_all = "true";
}

if (!defined($opts{d}) && !defined($opts{t})) {
	print "\nError: a database name or table name is required.\n\n";
	HELP_MESSAGE();
}

if (!defined($opts{d}) && defined($opts{t})) {
	print "\nError: a database name must accompany a table name.\n\n";
	HELP_MESSAGE();
}

format LINE = 
@<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<< @*
$name, $results
.

if (defined($opts{d})) {
	#-rw-------  1 pgsql  pgsql  4 Dec 14 05:00 /pg90/data/base/753229/PG_VERSION
	$name 			= $arg_database;
	$data_source  	= "$driver:dbname=$default_db";
	$dbh 			= DBI->connect($data_source, $username, $password) or die $DBI::errstr;
	#$query			= "SELECT oid FROM pg_database WHERE datname = ?;";
	#get_results($dbh, $query, $arg_database, "database");
	get_results($dbh, $arg_database, undef, "database");
	if (defined($arg_all)) {
		$data_source  	= "$driver:dbname=$arg_database";
		if (defined($opts{a})) {
			$query			= "SELECT relname FROM pg_class WHERE relkind = 'r' ORDER BY relname;";
		} else {
			$query			= "SELECT relname FROM pg_class WHERE relkind = 'r' AND relhaspkey='t' ORDER BY relname;";
		}
		my $dbh 		= DBI->connect($data_source, $username, $password) or die $DBI::errstr;
		my $sth 		= $dbh->prepare($query);
		$sth->execute();
		while (my $array_ref = $sth->fetchrow_arrayref) {
			my $relname = $array_ref->[0];
			$name 		= $arg_database . "." . $relname;
			#$query		= "SELECT oid FROM pg_class WHERE relname = ? AND relkind = 'r';";
			get_results($dbh, $arg_database, $relname, "table");
		}
		$sth->finish;
	}
	$dbh->disconnect;
}

if (defined($opts{t})) {
	#-rw-------  1 pgsql  pgsql  25616384 Dec 15 03:39 /pg90/data/base/736895/738844
	$name 			= $arg_database . "." . $arg_table;
	$data_source  	= "$driver:dbname=$arg_database";
	$dbh 			= DBI->connect($data_source, $username, $password) or die $DBI::errstr;
	#$query			= "SELECT oid FROM pg_class WHERE relname = ? AND relkind = 'r';";
	get_results($dbh, $arg_database, $arg_table, "table");
	$dbh->disconnect;
}

exit(0);

sub get_results {
	#my ($dbh, $this_query, $this_arg, $this_type) = @_;
	my ($dbh, $this_database, $this_table, $this_type) = @_;
	my ($sth, $oid, $stat_path, $ls_cmd, $timestamp);

	#$sth 			= $dbh->prepare($this_query);
	#$sth->execute($this_arg);
	#$query_results 	= $sth->fetchrow_arrayref;
	#$sth->finish;

	#$oid 			= get_oid($this_type, $query_results->[0], $this_arg);
	if ($this_type eq "database") {
		$oid		= `sudo ./pgoid.pl -d $this_database`;
		chomp ($oid);
		$db_oid 	= $oid;
		$stat_path	= "$pgdata_path/$db_oid/PG_VERSION";
	} elsif ($this_type eq "table") {
		$oid		= `sudo ./pgoid.pl -d $arg_database -t $this_table`;
		chomp ($oid);
		$tb_oid 	= $oid;
		$stat_path	= "$pgdata_path/$db_oid/$tb_oid";
	} else {
		$~ 			= "LINE";
		$results 	= "ERROR: Unknown type '$this_type' specified!";
		write();
		exit;
	}

	if (! -e $stat_path) {
		if (defined($arg_all)) {
			$timestamp = "Not found";
		} else {
			$~ 			= "LINE";
			$results 	= "ERROR: PostgreSQL data path '$stat_path' does not exist!";
			write();
			exit;
		}
	} else {
		$timestamp = ctime((stat($stat_path))[9]);
	}
	do_printing($timestamp, $stat_path);
}

sub sanitize {
	my ($input) = @_;

	$input =~s /\s+//g;
	$input =~s /;//g;
	$input =~s /[^\w]+//g;

	return $input;
}

sub get_oid {
	my ($type, $query_results) = @_;
	my $oid;

	if ($query_results && $query_results =~ /^(\d+)$/) {
		$oid = $1;
	} else {
		$~ = "LINE";
		$results =  "ERROR: Could not determine oid for $type $name! Does $name exist?\n";
		write();
		exit;
	}

	return $oid;
}

sub do_printing {
	my ($timestamp, $path) = @_;

	$~ = "LINE";
	$results        = "$timestamp $path";
	
	write; 
}

sub clean_up {
    exit(0);
}
